### ISSUE TITLE
Unable to get Edit / New ComboBoxes with LookupEditor working correctly

### ISSUE THREAD

**Issue by tamanes** [volkanceylan/Serenity#2947]
Hi Serenity,

I have an entity with a row named TaskRow, and it contains two fields that are related to users. I show you one of those. I am trying to implement a LookupEditor in order to show Username instead of UserId. For my grid, I found something working, but not for the comboboxes of Edit and New dialog.

I am trying to do things exactly as in Northwind.OrderRow, that looks up EmployeeRow items and it's comboboxes work ok. I cannot figure out what I'm doing different. As you can see in my code below, I tried also a million other combinations.

The one that works best is what I have marked //ORI, that uses [Expression()] instead of [Origin()]. I see usernames in my comboboxes, then. But: changes are not saved, and on opening the dialogs the combobox is at "select--" position, not at current selected item.

The current incommented code gives me an exception, already in SiteInitialization:
"ArgumentOutOfRangeException: Property 'UserExecutedByName' on row type 'TaskRow' has a [Origin] attribute, but its corresponding property 'UserExecutedByName' on row type 'UserRow' is not found!"_

I mainly seem to be unable to play with TextualField (from examples, I think to see it refers to my row, but seems to try lookup in UserRow) and Origin attribute, while also UserRow.RowFields LocalTextPrefix = "Administration.User" seems to maybe have to do with it.
I'm not sure how Serenity knows that I want UserRow.Username (just like I'm not sure, in the Northwind example, how it knows they want EmployeeRow.FullName).

Any ideas? Thanks in advance.
Jan Willem

    using XXX.Administration.Entities;

        //ORI:
        //[DisplayName("User Executed By"), NotNull, ForeignKey("[public].[Users]", "UserId"), LeftJoin("jUserExecutedBy"), TextualField("UserExecutedByUsername")]
        //EXPERI:
        //[DisplayName("User Executed By"), NotNull, ForeignKey(typeof(UserRow), "UserId"), LeftJoin("jUserExecutedBy")]
        //[LookupEditor(typeof(UserRow)), TextualField("UserExecutedByName")]
        //[LookupEditor(typeof(UserRow))]
        //[LookupEditor(typeof(UserRow)), TextualField("Username")]
        //EXPERI:
        //[DisplayName("User Executed By"), NotNull, ForeignKey("[public].[Users]", "UserId"), LeftJoin("jUserExecutedBy")]
        [DisplayName("User Executed By"), ForeignKey(typeof(UserRow)), LeftJoin("jUserExecutedBy")]
        [LookupEditor(typeof(UserRow)), TextualField("UserExecutedByName")]
        public virtual Int32? UserExecutedById
        {
            get { return Fields.UserExecutedById[this]; }
            set { Fields.UserExecutedById[this] = value; }
        }

        //ORI:
        //[DisplayName("User Executed By"), Expression("jUserExecutedBy.[Username]")]
        //[LookupEditor(typeof(AppreziSereneMVC.Administration.Entities.UserRow))]
        //EXPERI:
        //[Expression("jUserExecutedBy.Username"), DisplayName("User Executed By")]
        //public String Username
        //public String DisplayName
        [Origin("jUserExecutedBy"), DisplayName("User Executed By")]
        public String UserExecutedByName
        {
            get { return Fields.UserExecutedByName[this]; }
            set { Fields.UserExecutedByName[this] = value; }
        }

This is Northwind.OrderRow code:

        [DisplayName("Employee"), ForeignKey(typeof(EmployeeRow)), LeftJoin("e")]
        [LookupEditor(typeof(EmployeeRow)), TextualField("EmployeeFullName")]
        public Int32? EmployeeID
        {
            get { return Fields.EmployeeID[this]; }
            set { Fields.EmployeeID[this] = value; }
        }

        [Origin("e"), DisplayName("Employee")]
        public String EmployeeFullName
        {
            get { return Fields.EmployeeFullName[this]; }
            set { Fields.EmployeeFullName[this] = value; }
        }

**Comment by davidzhang9990** [comment]
check your UserRow.cs  ,find IName field and set it "Username".
try it.

**Comment by davidzhang9990** [comment]
![image](https://user-images.githubusercontent.com/23019715/32987686-ad7e395c-cd2d-11e7-9971-626ee3378968.png)

**Comment by tamanes** [comment]
Hi davidzhang9990, thank you! But your code is already in place, in UserRow, it's the UserRow from Serene example, only extended with some extra fields.

But anyway I learn something more about the mechanism, I now see how the Northwind OrderRow gets EmployeeRow.FullName from the lookup.

**Comment by Zahar661** [comment]
TaskRow and UserRow in one base?

**Comment by tamanes** [comment]
Hi Zahar661, thanks for your help.
What do you mean, one base? I see that UserRow derives from LoggingRow, which derives from Row. While TaskRow derives from Row. Is that a problem?
(For fun, I now tried to derive TaskRow from LoggingRow also. No, same exception in SiteInitialization).

**Comment by Zahar661** [comment]
in one BD?

**Comment by tamanes** [comment]
Hi Zahar661, I'm sorry, what is a BD?
regards, Jan Willem

**Comment by Zahar661** [comment]
Database. If different, then the Join does not work.  Table Users in DB "MyProject_Default_v1"

**Comment by tamanes** [comment]
I see. Hmmmyes, both UserRow and TaskRow have [ConnectionKey("Default")], it's the Default database.
Now if I make that intentionally wrong, I give only UserRow [ConnectionKey("Northwind")], then the exception is the same ... it seems like SiteInitialization does not yet look in databases, maybe I'm wrong ... I wonder if UserRow has something special, different from Nortwind.EmployeeRow.

**Comment by Zahar661** [comment]
UserExecutedByName rename UserExecutedByUsername  or UserExecutedByDisplayName

**Comment by tamanes** [comment]
Hi Zahar661,

Thanks a lot for your efforts! and I think you are very clever to detect that this matters! I renamed to
UserExecutedByUsername and .... no more exception! But I'm confused, so for [Origin] it is obligatory that the propertyname ends with the propertyname in the looked up row? I thought "TextualField" just indicates the name of the property in my row.

(Then in order to try resolve other problem, I renamed UserExecutedById to UserExecutedByUserId (so that it ends on "UserId"), also in the Fields, also in the database. Guess what: the very same exception was back, now with property 'UserExecutedByUserName'!)

But actually, now with exception gone, my original problem turned out very different from what I thought: I had incorrectly thought that in xxxForm.cs, I had to take String UserExecutedByUserName instead of Int32 UserExecutedById !

Now that I have that, I can abandon [Origin] and use:

	[DisplayName("User Executed By"), NotNull, ForeignKey(typeof(UserRow), "UserId"), LeftJoin("jUserExecutedBy"), TextualField("UserExecutedByName")]
        [LookupEditor(typeof(UserRow))]
        public virtual Int32? UserExecutedById
        {
        }

and

        [DisplayName("User Executed By"), Expression("jUserExecutedBy.[Username]")]
        public String UserExecutedByName
        {
        }

(renamed back to UserExecutedByName)
and it works! So I wonder: what is [Origin] for (not in my serenity-guide) ?

best regards & thanks, Jan Willem
